libcalico-go/lib/epstatusfile/status_file_watcher.go

```diff
@@ -0,0 +1,228 @@
+
+package epstatusfile
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+)
+
+// FileWatcher monitors a directory and reports events (creation, update and deletion) on file updates.
+// It tries to use fsnotify to watch the directory. If fsnotify fails, it switches to polling mode.
+
+type Callbacks struct {
+	OnFileCreation func(fileName string)
+	OnFileUpdate   func(fileName string)
+	OnFileDeletion func(fileName string)
+}
+
+type FileWatcher struct {
+	dir                 string
+	lastState           map[string]os.FileInfo
+	mu                  sync.Mutex
+	stopChan            chan struct{}
+	fsWatcher           *fsnotify.Watcher
+	pollIntervalSeconds int
+
+	callbacks Callbacks
+}
+
+// NewWatcher creates a new Watcher instance.
+func NewFileWatcher(dir string, pollIntervalSeconds int) *FileWatcher {
+	return &FileWatcher{
+		dir:                 dir,
+		lastState:           make(map[string]os.FileInfo),
+		stopChan:            make(chan struct{}),
+		pollIntervalSeconds: pollIntervalSeconds,
+	}
+}
+
+func (w *FileWatcher) SetCallbacks(callbacks Callbacks) {
+	w.callbacks = callbacks
+}
+
+func (w *FileWatcher) Start() {
+	go w.runWatcher()
+}
+
+// Start begins watching the directory.
+func (w *FileWatcher) runWatcher() {
+	var err error
+
+	// Get current state of the directory and emit initial events.
+	w.checkForChanges()
+
+	w.fsWatcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		log.WithError(err).Error("Error initializing fsnotify. Falling back to polling.")
+		go w.pollDirectory()
+		return
+	}
+	defer w.fsWatcher.Close()
+
+	err = w.fsWatcher.Add(w.dir)
+	if err != nil {
+		log.WithError(err).Error("Error adding directory to fsnotify. Falling back to polling.")
+		go w.pollDirectory()
+		return
+	}
+
+	log.WithField("dir", w.dir).Info("Started watching directory.")
+
+	// Listen for events
+	for {
+		select {
+		case event, ok := <-w.fsWatcher.Events:
+			if !ok {
+				// Stop if channel is closed.
+				return
+			}
+			log.WithFields(log.Fields{
+				"file": event.Name,
+				"op":   event.Op,
+			}).Debug("received file events")
+
+			filePath := event.Name
+			if event.Op == fsnotify.Remove {
+				w.callbacks.OnFileDeletion(filePath)
+				delete(w.lastState, filePath)
+			} else {
+				fileInfo, err := os.Stat(filePath)
+				if err != nil {
+					log.WithError(err).Error("Failed to get file info on a fsnotify event.")
+				} else if !fileInfo.IsDir() {
+					if event.Op == fsnotify.Create {
+						w.lastState[filePath] = fileInfo
+						w.callbacks.OnFileCreation(filePath)
+					}
+					if event.Op == fsnotify.Write {
+						w.lastState[filePath] = fileInfo
+						w.callbacks.OnFileUpdate(filePath)
+					}
+				}
+			}
+
+		case err, ok := <-w.fsWatcher.Errors:
+			if !ok {
+				// Stop if channel is closed.
+				return
+			}
+			log.WithError(err).Error("fsnotify error. Falling back to polling.")
+			go w.pollDirectory()
```

Comment:
Seems inconsistent to start polling here but not if the other channel is closed.  I'm not keen on the decision to start polling being spread around throughout the function.  I'd probably have an outer retry loop for recreating the watch and do polling whenever we don't have an active watch.  You probably also need to do a poll at start of day after establishing the watch to make sure you didn't miss anything?

think it might be clearer not to move polling to a separate goroutine, just do the polling on this goroutine, then you can't get confused about whether a poll is running.